base/timing.jl

```diff
@@ -279,6 +279,31 @@ macro allocated(ex)
     end
 end
 
+"""
+    @compiletime
+
+A macro to evaluate an expression, discarding the resulting value, instead returning the
+total time in seconds of code generation during evaluation of the expression.
+
+See also [`@time`](@ref), [`@timev`](@ref), [`@timed`](@ref),
+and [`@elapsed`](@ref).
+
+```julia-repl
+julia> x = rand(10,10);
+
+julia> @compiletime x * x
+0.54208595
+```
+"""
+macro compiletime(ex)
+    quote
+        while false; end # compiler heuristic: compile this block (alter this if the heuristic changes)
+        local t0 = cumulative_compile_time_ns()
```

Comment:
Note that this number is inaccurate since it ignores the compile time of the macro itself.

It was OK as an additional info in `@time` output since the number is to be compared to the other outputs none of which includes the actual compile time. However, this is not the case when the compile time is the only output. People will think `@compiletime function_that_take_a_minute_to_compile()` has no compile time. At a minimum, the `while false; end` should be removed and even that is not accurate since the user code can still trigger a compilation.

Reply:
Thanks for the additional review. I'm happy to revert the last commit. Just wanted to check that reviewers had considered it, given it came up in discussion.

For the sake of clarity though, I'm not sure I understand a few of your points. 

> inaccurate since it excludes the compile time of the macro itself

Why would we want to include the compile time of the macro in this timing? It's this way by design. 

> none of which includes the actual compile time. 

The total of `@time` includes any compile time, that's the motivation for this PR. To highlight when compile time is present, as it usually dominates. 

> people will think ... has no compile time

If the function you describe does take a minute to compile, the first time it's `@compiletime`-ed, it would return 60 seconds. Though, from then onwards it would return 0, which could be confusing. Is that what you mean?


Again, I'm not pushing for this additional macro too hard. I think the main benefit is the addition to `@time`

Reply:
> Why would we want to include the compile time of the macro in this timing? It's this way by design.

Because that compilation might include the compilation of the function in question. See below.

> The total of `@time` includes any compile time, that's the motivation for this PR. To highlight when compile time is present, as it usually dominates.

No it doesn't. It sometimes include the compilation of the function on the first use which is why adding compilation time is useful. However, even on the first execution, the compile time may not be included, and that's exactly what that comment on the first line of the macro expansion means. Again, see below.

> If the function you describe does take a minute to compile, the first time it's `@compiletime`-ed, it would return 60 seconds. Though, from then onwards it would return 0, which could be confusing. Is that what you mean?

Nope. See the example below (and change the 1000 to 10000 if you dare.... that gives a longer than 1min compilation time for me for sure, though to be fare, it was still inferring after 5 min so I'm not sure how long each phase takes).

```julia
julia> @eval function f(a)
           $(quote $([:(a = a + a * 10 + (a > 0 ? $i : 0.2)) for i in 1:1000]...) end)
           $(quote $([:(a = a + a * 10 + (a > 0 ? $i : 0.2)) for i in 1:1000]...) end)
           return a
       end
f (generic function with 1 method)

julia> @time f(1)
  0.000000 seconds
Inf
```

The point I used a no argument function is that that function is guaranteed to be inferred and compiled based on the current implementation.

Reply:
I see. So there may be more compilation going on related to the function under test that none of this timing tooling is catching, so it's only relevant to report when the time returned by `@time` contains any compiled time. (and in that case it doesn't make sense to report an absolute time given it's not a total, so the % is adequate as a highlight)

Makes sense, thanks for the explanation. I'll revert

Reply:
Thanks both for the above conversation! I was really concerned about this, since even _before_ this change I was recently misled to find that code I thought took 30 seconds to compile actually too _100 seconds_ to compile, but most of the compilation time was spent in the top-level inference, exactly as in @yuyichao's example here. 😮 

I was worried about this new `% compilation time` metric (which i'm super thrilled about!!!) making it even easier to be confused.

So @vchuravy and I took a stab at reporting the top-level compilation time as well, here:
https://github.com/JuliaLang/julia/pull/37938